sdks/python/apache_beam/io/datastore/v1/datastoreio.py

```diff
@@ -0,0 +1,289 @@
+      try:
+        query_splits = query_splitter.get_splits(
+            self._datastore, query, estimated_num_splits,
+            helper.make_partition(self._project, self._datastore_namespace))
+      except Exception:
+        logging.warning("Unable to parallelize the given query: %s", query,
+                        exc_info=True)
+        query_splits = [(key, query)]
+
+      sharded_query_splits = []
+      for split_query in query_splits:
+        sharded_query_splits.append((key, split_query))
+        key += 1
+
+      return sharded_query_splits
+
+    def display_data(self):
+      disp_data = {'project': self._project,
+                   'query': str(self._query),
+                   'num_splits': self._num_splits}
+
+      if self._datastore_namespace is not None:
+        disp_data['namespace'] = self._datastore_namespace
+
+      return disp_data
+
+  class ReadFn(DoFn):
+    """A DoFn that reads entities from Cloud Datastore, for a given query."""
+
+    def __init__(self, project, namespace=None):
+      super(ReadFromDatastore.ReadFn, self).__init__()
+      self._project = project
+      # using _namespace conflicts with DisplayData._namespace
+      self._datastore_namespace = namespace
+      self._datastore = None
+
+    def start_bundle(self, context):
+      self._datastore = helper.get_datastore(self._project)
+
+    def process(self, p_context, *args, **kwargs):
+      query = p_context.element
+      # Returns an iterator of entities that reads in batches.
+      entities = helper.fetch_entities(self._project, self._datastore_namespace,
+                                       query, self._datastore)
+      return entities
+
+    def display_data(self):
+      disp_data = {'project': self._project}
+
+      if self._datastore_namespace is not None:
+        disp_data['namespace'] = self._datastore_namespace
+
+      return disp_data
+
+  @staticmethod
+  def query_latest_statistics_timestamp(project, namespace, datastore):
+    """Fetches the latest timestamp of statistics from Cloud Datastore.
+
+    Cloud Datastore system tables with statistics are periodically updated.
+    This method fethes the latest timestamp (in microseconds) of statistics
+    update using the `__Stat_Total__` table.
+    """
+    query = helper.make_latest_timestamp_query(namespace)
+    req = helper.make_request(project, namespace, query)
+    resp = datastore.run_query(req)
+    if len(resp.batch.entity_results) == 0:
+      raise RuntimeError("Datastore total statistics unavailable.")
+
+    entity = resp.batch.entity_results[0].entity
+    return datastore_helper.micros_from_timestamp(
+        entity.properties['timestamp'].timestamp_value)
+
+  @staticmethod
+  def get_estimated_size_bytes(project, namespace, query, datastore):
+    """Get the estimated size of the data returned by the given query.
+
+    Cloud Datastore provides no way to get a good estimate of how large the
```

Comment:
It is difficult to parse this comment. Is it mentioning the problem and the work around in the same sentence? Could you simplify it?